lxd/instance/drivers/driver_qemu.go

```diff
@@ -1559,115 +1559,242 @@ func (vm *qemu) generateQemuConfigFile(bus string, devConfs []*deviceConfig.RunC
 		return "", err
 	}
 
-	// Now add the dynamic parts of the config.
-	err = qemuSerial.Execute(sb, map[string]interface{}{
-		"architecture":     vm.architectureName,
-		"ringbufSizeBytes": qmp.RingbufSize,
-	})
+	err = vm.addCPUConfig(sb)
 	if err != nil {
 		return "", err
 	}
 
-	err = qemuSCSI.Execute(sb, map[string]interface{}{
-		"architecture": vm.architectureName,
-	})
+	err = vm.addMemoryConfig(sb)
 	if err != nil {
 		return "", err
 	}
 
-	err = qemuBalloon.Execute(sb, map[string]interface{}{
+	err = qemuDriveFirmware.Execute(sb, map[string]interface{}{
 		"architecture": vm.architectureName,
+		"roPath":       filepath.Join(vm.ovmfPath(), "OVMF_CODE.fd"),
+		"nvramPath":    vm.nvramPath(),
 	})
 	if err != nil {
 		return "", err
 	}
 
-	err = qemuRNG.Execute(sb, map[string]interface{}{
-		"architecture": vm.architectureName,
+	err = qemuControlSocket.Execute(sb, map[string]interface{}{
+		"path": vm.monitorPath(),
 	})
 	if err != nil {
 		return "", err
 	}
 
-	err = vm.addMemoryConfig(sb)
-	if err != nil {
-		return "", err
+	// getBusAddress is used to create any needed root ports and provide
+	// the bus and address that should be used by a device. It supports
+	// automatically setting up multi-function devices and optimize their use.
+	portNum := 0
+	devNum := 5
+
+	type entry struct {
+		bridgeDev int // Device number on the root bridge.
+		bridgeFn  int // Function number on the root bridge.
+
+		dev string // Existing device name.
+		fn  int    // Function number on the existing device.
 	}
+	entries := map[string]*entry{}
 
-	err = vm.addCPUConfig(sb)
+	var rootPort *entry
+	getRootPort := func() *entry {
+		if rootPort == nil {
+			rootPort = &entry{
+				bridgeDev: devNum,
+			}
+			devNum++
+		} else {
+			if rootPort.bridgeFn == 7 {
+				rootPort.bridgeFn = 0
+				rootPort.bridgeDev = devNum
+				devNum++
+			} else {
+				rootPort.bridgeFn++
+			}
+		}
+
+		return rootPort
+	}
+
+	getBusAddr := func(group string) (string, string, bool) {
```

Comment:
@stgraber I wonder if it would be clearer to change `group` arg its own int/string type and use constants to indicate the groups, then we can add comments next to constant definitions explaining why they get their own groups. And the empty one can be called `NONE` or similar.

Reply:
I guess we can do that when we move to a separate struct.